core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AbstractMaterializeTableScan.java

```diff
@@ -0,0 +1,217 @@
+import io.trino.sql.ir.Expression;
+import io.trino.sql.ir.Row;
+import io.trino.sql.planner.Symbol;
+import io.trino.sql.planner.iterative.Rule;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.TableScanNode;
+
+import java.io.IOException;
+import java.io.InterruptedIOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeoutException;
+
+import static io.airlift.concurrent.MoreFutures.getFutureValue;
+import static io.trino.SystemSessionProperties.getMaterializeTableMaxActualRowCount;
+import static io.trino.SystemSessionProperties.getMaterializeTableMaxEstimatedRowCount;
+import static io.trino.SystemSessionProperties.getMaterializeTableTimeout;
+import static io.trino.SystemSessionProperties.isMaterializeTableEnabled;
+import static io.trino.spi.type.TypeUtils.readNativeValue;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public abstract sealed class AbstractMaterializeTableScan<T extends PlanNode>
+        implements Rule<T>
+        permits MaterializeFilteredTableScan, MaterializeTableScan
+{
+    private static final int MAX_SPLITS = 10_000;
+
+    protected final Logger log = Logger.get(getClass());
+    protected final PlannerContext plannerContext;
+    private final SplitManager splitManager;
+    private final PageSourceManager pageSourceManager;
+    private final ExecutorService executor;
+    private final HostAddress currentNode;
+
+    protected AbstractMaterializeTableScan(
+            PlannerContext plannerContext,
+            SplitManager splitManager,
+            PageSourceManager pageSourceManager,
+            InternalNodeManager nodeManager,
+            ExecutorService executor)
+    {
+        this.plannerContext = requireNonNull(plannerContext, "plannerContext is null");
+        this.splitManager = requireNonNull(splitManager, "splitManager is null");
+        this.pageSourceManager = requireNonNull(pageSourceManager, "pageSourceManager is null");
+        this.executor = requireNonNull(executor, "executor is null");
+        this.currentNode = nodeManager.getCurrentNode().getHostAndPort();
+    }
+
+    @Override
+    public final boolean isEnabled(Session session)
+    {
+        return isMaterializeTableEnabled(session) &&
+                getMaterializeTableMaxEstimatedRowCount(session) > 0 &&
+                getMaterializeTableMaxActualRowCount(session) > 0 &&
+                getMaterializeTableTimeout(session).toMillis() > 0;
+    }
+
+    @Override
+    public final Result apply(T node, Captures captures, Context context)
+    {
+        int maxRows = getMaterializeTableMaxEstimatedRowCount(context.getSession());
+        PlanNodeStatsEstimate stats = context.getStatsProvider().getStats(node);
```

Comment:
I think this is potentially problematic for filtered tables.
The scan materialization is only going to benefit from the part of the filter that is enforced by the connector page or split source, but the estimate is going to include the filtering from the entire predicate.
Maybe this should just be estimate of output of TableScan as that should include any filtering that we gain from `enforcedConstraint`.

Reply:
That's a good point. How do I fetch the stats for that?